### DISCUSSION TITLE
Mapping UserProfile attribute to claim in token

### DISCUSSION THREAD

**Discussion by scott-mac** [keycloak/keycloak#23587] [Q&A] [answered]
On version 22, we activated the UserProfile option and added several custom attributes that we wish to record on our users. 
These work well and we can update them via the Admin console or using the REST API. 

We need to be able to expose one of them as a claim in the token though and I don't 'think' there is an obvious way to do this unless I'm missing something. I can go to a user's details and add a custom attribute there and then map that to the token and it shows up but it doesn't seem to work for the User Profile attributes. Is there a way to accomplish this in the console? Or is this a scenario where we will need to write a custom protocol mapper? And if the latter, curious if anyone has tried building such a thing. 

thank you for your help with this.

**Comment by jbman** [comment, marked as answer]
Did you add a user attribute mapper to a default scope of the client?
You can add a user attribute mapper to the dedicated client scope:
<img src="https://github.com/keycloak/keycloak/assets/789496/81173f82-c59b-4123-a4a5-9fb2f55e0805" width=60% height=60%>

Example of a mapper for attribute "department" to claim "department":
<img src="https://github.com/keycloak/keycloak/assets/789496/650c103f-f0df-4cd0-b714-8f21b1c128b3" width=50% height=50%>

The evaluate feature then shows the claim in the token:
<img src="https://github.com/keycloak/keycloak/assets/789496/b147626e-4b29-4aa0-b0a9-18e4daee456e" width=60% height=60%>

**Comment by scott-mac** [comment]
Thanks for the reply jbman and the detailed walkthru. 
The baffling part is that I followed these exact steps previously but wasn't seeing the attribute appearing in the token. The only thing different is that the docker instance was rebooted since, i'm wondering if something was hung up or cached and the reboot fixed it? 

in any event, thanks for the help, I appreciate it!

**Reply by asv70** [reply]
you need set Realm Settings -> General -> Unmanaged Attributes -> Enabled to attribute appearing in the token.
